### PR TITLE
Image view

### PR DESCRIPTION
**Pull request by danicadombek**
Story ac`s:
- Click of an image open a new screen with image-preview/ detailview
- Button to get back to the image-list
- Component w/ storybook and test
- Update App with component and logic

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -10,6 +10,7 @@
     "prop-types": "^15.7.2",
     "react": "^17.0.2",
     "react-dom": "^17.0.2",
+    "react-router-dom": "^5.2.0",
```
Comment (badabam) [inline package.json:13]: Ah, you are going to use react router!

File: src/App.js [hunk 2]
```diff
@@ -26,7 +31,26 @@ export default function App() {
       }}
     >
       <Header>Viary</Header>
-      <MemoriesPage memories={listOfMemories} />
+      {currentPage === 'memories' && (
```
Comment (badabam) [inline src/App.js:35]: What happend to react router?

File: src/components/Button.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import styled from 'styled-components/macro'
+import PropTypes from 'prop-types'
+
+Button.propTypes = {
+  onClick: PropTypes.func,
+  children: PropTypes.node,
+}
```
Comment (badabam) [inline src/components/Button.js]: Nice propTypes!

File: src/components/Button.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+const ButtonStyle = styled.button`
+  border-radius: 50px;
+  background: #e4eaeb;
+  box-shadow: 25px 25px 33px #baaba4;
+  max-width: auto;
+  font-size: 18px;
+  font-weight: semi-bold;
+`
```
Comment (badabam) [inline src/components/Button.js]:
```suggestion
export default styled.button`
  border-radius: 50px;
  background: #e4eaeb;
  box-shadow: 25px 25px 33px #baaba4;
  max-width: auto;
  font-size: 18px;
  font-weight: semi-bold;
`
```

File: src/components/Button.spec.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import { render, screen } from '@testing-library/react'
+import userEvent from '@testing-library/user-event'
+import Button from './Button'
+
+describe('Button', () => {
+  it('contains a text', () => {
+    render(<Button onClick={() => {}}>Click me</Button>)
+
+    const button = screen.getByRole('button', { name: 'Click me' })
+    expect(button).toBeInTheDocument()
```
Comment (badabam) [inline src/components/Button.spec.js:10]: Good!

File: src/components/Button.spec.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  })
+
+  it('calls onClick', async () => {
+    const handleClick = jest.fn()
+    render(<Button onClick={handleClick}>Click me</Button>)
+
+    const button = screen.getByRole('button', { name: 'Click me' })
+    userEvent.click(button)
+
+    expect(handleClick).toHaveBeenCalledTimes(1)
```
Comment (badabam) [inline src/components/Button.spec.js:20]: Another good test!

File: src/components/Header.js [hunk 7]
```diff
@@ -7,9 +7,9 @@ Header.propTypes = {
 
 export default function Header({ children }) {
   return (
-    <Background>
+    <div>
```
Comment (badabam) [inline src/components/Header.js]: You don't really need this div, do you?

File: src/components/MemoryItem.js [hunk 8]
```diff
@@ -2,39 +2,33 @@ import styled from 'styled-components/macro'
 import PropTypes from 'prop-types'
 
 MemoryItem.propTypes = {
+  onDetail: PropTypes.func.isRequired,
```
Comment (badabam) [inline src/components/MemoryItem.js:5]: Nice, with `isRequired`!

File: src/components/MemoryItem.js [hunk 9]
```diff
@@ -2,39 +2,33 @@ import styled from 'styled-components/macro'
...
+export default function MemoryItem({ image, subtitle, onDetail }) {
   return (
-    <ListWrapper>
-      <li>
-        <Image src={image} alt="" width="100%" />
-        <span>{subtitle}</span>
-      </li>
-    </ListWrapper>
+    <Wrapper onClick={() => onDetail(image)}>
+      <Image src={image} alt="" width="100%" />
```
Comment (badabam) [inline src/components/MemoryItem.js]: An image should always get both, width and height attributes, but always without any unit. It's only to give the browser the expected aspect ratio. Please have a look at the Article that Mel shared today in _hh-web-21-3_

File: src/components/MemoryItem.js [hunk 10]
```diff
@@ -2,39 +2,33 @@ import styled from 'styled-components/macro'
...
-    display: grid;
-    align-items: center;
-    gap: 5px;
-    padding: 5px;
-    text-align: center;
-    color: #230707;
-    box-shadow: 2px 2px 4px grey;
-  }
+  text-align: center;
+  color: #230707;
```
Comment (badabam) [inline src/components/MemoryItem.js:26]: Please consider adding css variables in your GlobalStyle (in a `:root` selector).

File: src/pages/DetailMemoryPage.js [hunk 11]
```diff
@@ -0,0 +1,68 @@
+import styled from 'styled-components/macro'
+import Button from '../components/Button'
+import PropTypes from 'prop-types'
+import MemoriesIcon from '../assets/images/icons/galery.png'
+
+DetailMemoryPage.propTypes = {
+  image: PropTypes.node,
```
Comment (badabam) [inline src/pages/DetailMemoryPage.js:6]: This should be a string (as it is a url).

File: src/pages/DetailMemoryPage.js [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+  onNavigate: PropTypes.func.isRequired,
+}
+
+export default function DetailMemoryPage({ image, onNavigate }) {
+  return (
+    <Wrapper>
+      <ImageDetail>
+        <Image src={image} alt="Memory 1" width="100%" />
```
Comment (badabam) [inline src/pages/DetailMemoryPage.js:14]: the alt attribute should also be a prop, so you can have individual alt texts.

File: src/pages/DetailMemoryPage.js [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+      </ImageDetail>
+      <nav>
+        <ButtonStyle onClick={onNavigate}>
+          <Icon src={MemoriesIcon} alt="" />
+          <span>Back to memories</span>
+        </ButtonStyle>
```
Comment (badabam) [inline src/pages/DetailMemoryPage.js]: Please consider making this a `BackButton` component.

File: src/pages/DetailMemoryPage.js [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    </Wrapper>
+  )
+}
+
+const Wrapper = styled.section`
+  display: flex;
+  flex-direction: column;
+  align-items: center;
+  justify-content: space-between;
+  height: 86.9vh;
```
Comment (badabam) [inline src/pages/DetailMemoryPage.js]: That's an oddly specific height. Please check if it fits your demands on different browser (I.e. device) sizes.

File: src/pages/DetailMemoryPage.spec.js [hunk 15]
```diff
@@ -0,0 +1,24 @@
+import { render, screen } from '@testing-library/react'
+import userEvent from '@testing-library/user-event'
+import DetailMemoryPage from './DetailMemoryPage'
+
+describe('DetailMemoryPage', () => {
+  it('renders an image, an icon and a button', () => {
```
Comment (badabam) [inline src/pages/DetailMemoryPage.spec.js]: Better would be "two images" as that's what your test below does.

File: src/pages/DetailMemoryPage.spec.js [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+    expect(button).toBeInTheDocument()
+  })
+
+  it('calls onNavigate to another page', () => {
+    const onNavigate = jest.fn()
+    render(<DetailMemoryPage onClick={onNavigate} />)
+
+    const button = screen.getByRole('button')
+    userEvent.click(button)
+    expect(onNavigate).toBeCalledTimes(0)
```
Comment (badabam) [inline src/pages/DetailMemoryPage.spec.js]: Zero (`0`) times? Looks like an error.

File: src/pages/MemoriesPage.js [hunk 17]
```diff
@@ -3,22 +3,25 @@ import MemoryItem from '../components/MemoryItem'
...
     <Wrapper>
       <h2>Your memories</h2>
       <ListWrapper>
         {memories.map(memory => (
-          <MemoryItem
-            key={memory.image}
-            image={memory.image}
-            subtitle={memory.subtitle}
-          />
+          <li key={memory.image}>
```
Comment (badabam) [inline src/pages/MemoriesPage.js:18]: Nice! As the images are unique, this is truly a valid key.

**Review by badabam — CHANGES_REQUESTED** [review]
Even though I made quite some comments, I am happy about your achievements in this PullRequest. Keep on the good work! 👍

File: src/pages/DetailMemoryPage.spec.js (hunk 16 above)
Comment (danicadombek) [reply src/pages/DetailMemoryPage.spec.js]: With (1) times I get an error- I don't understand this behaviour. What could be the reason?

File: src/App.js (hunk 2 above)
Comment (danicadombek) [reply src/App.js:35]: I tried to make the back-button/ page-change with the react router. It felt strange to put it in the DetailMemoryPage and it only worked with a reload, so I changed the way. I read the using in undersides is possible, could this be another way? (Later I will use it anyway).

File: src/pages/DetailMemoryPage.spec.js (hunk 16 above)
Comment (badabam) [reply src/pages/DetailMemoryPage.spec.js]: Maybe you don't click the correct button? Maybe onNavigate is the wrong prop name? It should be called, right?

File: src/pages/DetailMemoryPage.js (hunk 11 above)
Comment (danicadombek) [reply src/pages/DetailMemoryPage.js:6]: It`s a local image-file.

File: src/pages/DetailMemoryPage.js (hunk 14 above)
Comment (danicadombek) [reply src/pages/DetailMemoryPage.js]: Looks also good on chrome with different mobile-sizes.

File: src/components/BackButton.spec.js [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+  })
+
+  it('calls onClick', () => {
+    const onNavigate = jest.fn()
+    render(<BackButton onClick={onNavigate}>Back</BackButton>)
+
+    const button = screen.getByRole('button')
+    userEvent.click(button)
+
+    expect(onNavigate).toHaveBeenCalledTimes(1)
```
Comment (badabam) [inline src/components/BackButton.spec.js:21]: Happy to see that this test works now. Could you please share, what was the reason for this not working in the first version?

File: src/components/Button.js [hunk 19]
```diff
@@ -1,7 +1,7 @@
 import styled from 'styled-components/macro'
 import PropTypes from 'prop-types'
 
-Button.propTypes = {
+styled.button.propTypes = {
```
Comment (badabam) [inline src/components/Button.js:4]: Didn't know this works! I would have just deleted the propTypes as a styled-component doesn't have restrictions on what props you can use – as long as they are allowed on HTMLElements.

File: src/components/BackButton.js [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+    </ButtonStyle>
+  )
+}
+
+const ButtonStyle = styled(Button)`
+  display: flex;
+  justify-content: space-evenly;
+  align-items: center;
+  position: fixed;
+  left: 23px;
```
Comment (JGieseWeb) [inline src/components/BackButton.js]: Why not use relative size?  Like 1,4em :)

File: src/components/MemoryItem.js [hunk 21]
```diff
@@ -27,8 +27,9 @@ const Wrapper = styled.section`
 `
 
 const Image = styled.img`
-  border-radius: 20px;
   -webkit-box-shadow: 0px 10px 13px -7px #000000,
     5px 5px 15px 5px rgba(0, 0, 0, 0);
-  box-shadow: 0px 10px 13px -7px #000000, 5px 5px 15px 5px rgba(0, 0, 0, 0);
+  box-shadow: var(--img-shadow);
+  aspect-ratio: attr(width) / attr(height);
```
Comment (badabam) [inline src/components/MemoryItem.js]: You don't need to do this as modern browsers do it on their own, if you set width and height attributes for an image.

File: src/components/MemoryItem.js [hunk 22]
```diff
@@ -10,7 +10,7 @@ MemoryItem.propTypes = {
 export default function MemoryItem({ image, subtitle, onDetail }) {
   return (
     <Wrapper onClick={() => onDetail(image)}>
-      <Image src={image} alt="" width="100%" />
+      <Image src={image} alt="" />
```
Comment (badabam) [inline src/components/MemoryItem.js]: You should set width and height attributes to give the browser the possibility to calculate the aspect ratio. Beware: don't set any units, but just the values like that: e.g. `<Image src={image} alt="" width="200" height="100" />`

**Review by badabam — APPROVED** [review]
Very nice, I see a lot of improvements. Please consider my remarks and then 🚢 it!

File: src/components/BackButton.spec.js (hunk 18 above)
Comment (danicadombek) [reply src/components/BackButton.spec.js:21]: The correct Proptype was "onClick" and I have supplemented jest.fn() before the first render.

**Pull request merged**